### DISCUSSION TITLE
Exchange to exchange routing using header values

### DISCUSSION THREAD

**Discussion by Kim-Robin** [rabbitmq/rabbitmq-server#7840] [Questions] [answered]
I am trying to have header exchange to route to two different exchange based on header value. However, I was not able to route based on the header, but only with routingkey. Is there way to route message to other exchange based on header instead of routingkey?

**Comment by michaelklishin** [comment, marked as answer]
Exchange-to-exchange routing does not imply any particular exchange type. The headers exchange type does not seem to have anything specific to queue bindings, at least at first glance.

`exchange.bind` supports a map of optional arguments just like `queue.bind` does.

Please put together an executable example, I cannot suggest anything else based on the information provided.

**Reply by Kim-Robin** [reply]
I was trying it on management console(ui). I wanted to see if exchangeA can route to exchangeB and exchangeC based oh header.

For example, if message published to exchangeA has header “x-should-route-to-b” : “true”, I want them to route to exchangeB. If not route to exchangeC.

i just want to know this is possible since i was not able to make it work on rabbitMQ management console(ui)

**Reply by michaelklishin** [reply]
I cannot suggest much with a description like that.

Here's an example that uses Bunny, a headers exchange that routes to a fanout exchange, which routes to a queue. Only a subset of message is routed based on a randomly picked header value:

``` ruby
#!/usr/bin/env ruby

require "bunny"

c  = Bunny.new; c.start
ch = c.create_channel
ch.confirm_select

x1  = ch.headers("7840.headers", durable: true)
x2  = ch.fanout("7840.fanout", durable: true)

x2.bind(x1, arguments: {"os" => "linux", "cores" => 8, "x-match" => "any"})

ch.queue("7840.q1", durable: true).bind(x2, routing_key: ".*")

possible_values = (0..10).to_a
15.times do |i|
  rand(100).times { x1.publish("", persistent: true, headers: {"cores" => possible_values.sample}) }
end

ch.wait_for_confirms
ch.close
c.close

puts "Done"
```

**Reply by michaelklishin** [reply]
Roughly 10% of 1500 messages will end up in queue `7840.q1`.

**Reply by Kim-Robin** [reply]
Sorry for the lack of description. You still manage to understand my question. Based on your code(I am not ruby developer), I am able to send messages to x1 and matching header of “core” : ”8” messages to x2 which will end up in queue, right? Thank you very much. After setting “x-match” to “any-with-x” it started to route properly. Sorry again for poor description

**Reply by michaelklishin** [reply]
Happy that the example helped you figure it out!
